Bowling.java

```diff
@@ -116,28 +120,50 @@ private Frame getCurrentFrame() {
 		return currentFrame;
 	}
 	
-	private Optional<Frame> previousFrame() {
-		if (allFrames.size() > 1) {
-			int nextToLastIndex = allFrames.size() - 2;
+	private Optional<Frame> getFrameBefore(Frame frame) {
+		int currentIndex = allFrames.indexOf(frame);
+		int lastIndex = allFrames.size() - 1;
+
+		if (allFrames.size() > 1 && currentIndex == lastIndex) {
+			int nextToLastIndex = allFrames.indexOf(frame) - 1;
 			return Optional.of(allFrames.get(nextToLastIndex));
+		} else if (allFrames.size() > 2) {
+			int thirdToLastIndex = allFrames.size() - 3;
+			return Optional.of(allFrames.get(thirdToLastIndex));
 		}
 		
 		return Optional.empty();
 	}
 
+	private Boolean isTurkey() {
+		return currentFrame.isStrike() &&
+			   secondToLastFrame.isPresent() && 
+			   secondToLastFrame.get().isStrike() && 
+			   thirdToLastFrame.isPresent() && 
+			   thirdToLastFrame.get().isStrike();
+	}
+
+	private Boolean isSingleStrike() {
+		return secondToLastFrame.isPresent() && 
+		       secondToLastFrame.get().isStrike() && 
+		       !currentFrame.isStrike();
+	}
+
 	private void updateTotalPoints() {
-		Frame currentFrame = getCurrentFrame();
-		if (!currentFrame.isSpare() && !currentFrame.isStrike()) {
-			Integer currentFrameTotal = currentFrame.getFrameScore();
-			totalPoints += currentFrameTotal;
-		}
-		
-		Optional<Frame> previousFrame = previousFrame();
-		if (previousFrame.isPresent() && previousFrame.get().isSpare() ) {
-			totalPoints += previousFrame.get().getFrameScore() + getCurrentFrame().firstBowl();
-		} else if (previousFrame.isPresent() && previousFrame.get().isStrike()) {
-			totalPoints += previousFrame.get().getFrameScore() + getCurrentFrame().getFrameScore();
-		}
+		currentFrame = getCurrentFrame();
+		secondToLastFrame = getFrameBefore(currentFrame);
+		if (secondToLastFrame.isPresent()) 
+			thirdToLastFrame = getFrameBefore(secondToLastFrame.get());
+
+		if (!currentFrame.isSpare() && !currentFrame.isStrike())
+			totalPoints += currentFrame.getFrameScore();
+
+		if (secondToLastFrame.isPresent() && secondToLastFrame.get().isSpare())
+			totalPoints += sparePoints + getCurrentFrame().firstBowl();
+		else if (isTurkey())
+			totalPoints += strikePoints * 3;
+		else if (isSingleStrike())
+			totalPoints += strikePoints + getCurrentFrame().getFrameScore();
```

Comment:
I'm hoping we'll see some pattern of this that might simplify the logic, but it's not apparent to me now.